### ISSUE TITLE
Newton-CG throws error when Hessian is a `scipy.sparse` class

### ISSUE THREAD

**Issue by david-cortes** [scipy/scipy#8792]
In `scipy.optimize.minimize`, some methods make use of a user-provided Hessian matrix. These tend to be quite sparse, but currently, the methods that make use of a Hessian, such as Newton-CG, would fail if the output from the provided `hess` parameter is a sparse matrix from the classes in `scipy.sparse`.

Particularly for Newton-CG, line 1578 in optimize.py:
`curv = numpy.dot(psupi, Ap)`
Would return a matrix rather than a scalar, and subsequent lines would fail, although I'm not sure if altering the function to work with sparse hessians up to that line would still make the rest of the module work well.

**Comment by jeffrey-hokanson** [comment]
I would like to confirm this issue is still present in SciPy 1.11.4. I've setup a minimal working example [gist](https://gist.github.com/jeffrey-hokanson/aa9b8b712e769a0943694f5e8fbe2c7b), which yields the error message

```
Traceback (most recent call last):
  File "/Users/jeffreyh/SVN/SciPy/test_opt_sparse.py", line 34, in <module>
    res = minimize(
          ^^^^^^^^^
  File "/opt/homebrew/lib/python3.11/site-packages/scipy/optimize/_minimize.py", line 707, in minimize
    res = _minimize_newtoncg(fun, x0, args, jac, hess, hessp, callback,
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/homebrew/lib/python3.11/site-packages/scipy/optimize/_optimize.py", line 2100, in _minimize_newtoncg
    if 0 <= curv <= 3 * float64eps:
       ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/homebrew/lib/python3.11/site-packages/scipy/sparse/_compressed.py", line 337, in __ge__
    return self._inequality(other, operator.ge, '_ge_',
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/homebrew/lib/python3.11/site-packages/scipy/sparse/_compressed.py", line 288, in _inequality
    raise NotImplementedError(" >= and <= don't work with 0.")
NotImplementedError:  >= and <= don't work with 0.
```

**Comment by dschmitz89** [comment]
@jeffrey-hokanson I ran into something similar at my day job some time back. What I did was replace `<=x` for sparse matrix ops with `<np.nextafter(x, np.inf)`. Maybe something similar would work here.

**Comment by perimosocordiae** [comment]
I think the correct fix here needs to happen before the comparison, as `curv` should be a scalar float in any case.

I haven't investigated yet, but it might be helpful to replace the np.dot() call with the @ operator.

**Comment by jeffrey-hokanson** [comment]
@perimosocordiae I think what's happening is that when the Hessian is a sparse matrix, the [line computing the curvature](https://github.com/scipy/scipy/blob/3eb93f5c6d8ec66baed453c49cbeba5f781ba85d/scipy/optimize/_optimize.py#L2186), doesn't yield a float. This line computes `Ap = np.dot(A, psupi)`, which for a sparse `A` and dense vector `psupi`. For example, using the functions from my [gist](https://gist.github.com/jeffrey-hokanson/aa9b8b712e769a0943694f5e8fbe2c7b), for these types we get a `np.ndarray` 

```python
H = rosenbrock_hess(x0)
g = rosenbrock_grad(x0)
t = np.dot(H, g)
print(t.__class__)
print(t)
```

results in 
```terminal
<class 'numpy.ndarray'>
[<2x2 sparse matrix of type '<class 'numpy.float64'>'
 	with 4 stored elements in COOrdinate format>
 <2x2 sparse matrix of type '<class 'numpy.float64'>'
 	with 4 stored elements in COOrdinate format>       ]
```

This then is not comparable with a float, resulting in the error message.


I agree using `@` would be more modern. However, since `HessianUpdateStrategy` defines a `dot` method and not a `__matmul__`, using `dot` allows us to avoid the branching logic. Perhaps adding `__matmul__` to `HessianUpdateStrategy` as an alias of `dot` would be beneficial.

**Comment by dschmitz89** [comment]
> I think the correct fix here needs to happen before the comparison, as `curv` should be a scalar float in any case.
> 
> I haven't investigated yet, but it might be helpful to replace the np.dot() call with the @ operator.

Yes that's what I also noticed when I looked deeper into it. I implemented an (in my opinion) easy fix in #19785 . Would you have time to review that PR @perimosocordiae ?

**Comment by perimosocordiae** [comment]
Thanks @dschmitz89, your PR looks good to me.
